### PR TITLE
Expose Safe and Idempotent on Protobuf descriptors

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
This PR is split into two commits, one to stabilize the MethodDescriptors, and the other to update the generated code.    

PTAL @ejona86 

Updates https://github.com/grpc/grpc-java/issues/1775

### CODE REVIEW

File: compiler/src/java_plugin/cpp/java_generator.cpp [hunk 1]
```diff
@@ -365,6 +365,21 @@ static void PrintMethodFields(
     (*vars)["arg_in_id"] = to_string(2 * i);
     (*vars)["arg_out_id"] = to_string(2 * i + 1);
     (*vars)["method_name"] = method->name();
+    bool safe = method->options().idempotency_level()
+        == google::protobuf::MethodOptions_IdempotencyLevel_NO_SIDE_EFFECTS;
```
Comment (dapengzhang0) [inline compiler/src/java_plugin/cpp/java_generator.cpp]: Wonder if we can add a safe unary method in golden test.proto.

**Comment by carl-mastrangelo** [comment]
It looks like the API is okay to stabilize.    Can this PR proceed?

File: compiler/src/java_plugin/cpp/java_generator.cpp (hunk 1 above)
Comment (carl-mastrangelo) [reply compiler/src/java_plugin/cpp/java_generator.cpp]: Works for me.  I can get to it tonight, or do a followup PR if this one can move forward.

**Comment by ejona86** [comment]
Yes, it can proceed. Although @dapengzhang0 noticed that OkHttp is wrongly enabling use of safe methods.
https://github.com/grpc/grpc-java/blob/28323e2fb6a4966b520e3a19ca19f42bafc8f44c/okhttp/src/main/java/io/grpc/okhttp/OkHttpChannelBuilder.java#L136

**Comment by carl-mastrangelo** [comment]
>  OkHttp is wrongly enabling use of safe methods.

Fixed.

I added the Safe and Idempotent compiler cases, PTAL.

**Comment by carl-mastrangelo** [comment]
Friendly ping on this.

**Comment by carl-mastrangelo** [comment]
@dapengzhang0 @ejona86 Is there any chance this can make the branch cut?

**Comment by ejona86** [comment]
I do think this could make it before the branch cut.

File: compiler/src/test/proto/grpc/testing/compiler/test.proto [hunk 2]
```diff
@@ -70,6 +70,18 @@ service TestService {
   // An RPC method whose Java name collides with a keyword, and whose generated
   // method should have a '_' appended.
   rpc Import(stream StreamingInputCallRequest) returns (stream StreamingInputCallResponse);
+
+  // A unary call that is Safe.   Implementations may use the "GET" method
```
[the remark this answers is not in the document]
Comment (carl-mastrangelo) [reply compiler/src/test/proto/grpc/testing/compiler/test.proto]: Done.

File: compiler/src/java_plugin/cpp/java_generator.cpp [hunk 3]
```diff
@@ -411,6 +426,8 @@ static void PrintMethodFields(
         "            $MethodDescriptor$.<$input_type$, $output_type$>newBuilder()\n"
         "            .setType($MethodType$.$method_type$)\n"
         "            .setFullMethodName(generateFullMethodName(SERVICE_NAME, \"$method_name$\"))\n"
+        "            .setSafe($safe$)\n"
```
[the remark this answers is not in the document]
Comment (carl-mastrangelo) [reply compiler/src/java_plugin/cpp/java_generator.cpp]: Done.  That said, I think you should look at the comment I left on #6285.

**Comment by carl-mastrangelo** [comment]
ok to merge?

**Comment by ejona86** [comment]
@carl-mastrangelo, thank you!

**Pull request merged**